python/GafferImageUI/CatalogueUI.py

```diff
@@ -131,6 +135,81 @@
 
 )
 
+
+##########################################################################
+# Viewer hot-keys
+##########################################################################
+
+def addCatalogueHotkeys( editor ) :
+
+	if not isinstance( editor, GafferUI.Viewer ) :
+		return
+
+	editor.keyPressSignal().connect( _handleCatalogueHotkeys, scoped = False )
+
+def _handleCatalogueHotkeys( viewer, event ) :
+
+	# Up/Down arrows need to walk upstream of the viewer input and look for
+	# a Catalogue node and increment/decrement its active index
+
+	if event.key not in ( "Down", "Up" ) :
+		return False
+
+	if not isinstance( viewer.view(), GafferImageUI.ImageView ) :
+		return False
+
+	catalogue = __findCatalogue( viewer.view() )
+	if catalogue is None :
+		return False
+
+	__incrementImageIndex( catalogue, event.key )
+
+	return True
+
+def __findCatalogue( node ) :
+
+	if isinstance( node, GafferImage.Catalogue ) :
+		return node
+	else :
+		for inPlug in GafferImage.ImagePlug.InputRange( node ) :
+			upstreamPlug = inPlug.getInput()
+			if upstreamPlug is None :
+				continue
+			catalogue = __findCatalogue( upstreamPlug.node() )
+			if catalogue is not None :
+				return catalogue
+
+	return None
+
+def __incrementImageIndex( catalogue, direction ) :
+
+	# Match the UI's top-to-bottom order instead of 'up is a larger number'
+	increment = -1 if direction == "Up" else 1
+
+	# The Catalog UI re-orders images internally using metadata, rather than by
+	# shuffling plugs. As such, we can't just set imageIndex. We don't want to
+	# be poking into the specifics of how this works, so for now we re-use
+	# _ImagesPath as it knows all that logic.
+
+	images = catalogue["images"].children()
+	if len( images ) == 0 :
+		return
+
+	maxIndex = len( images ) - 1
+	orderedImages = _ImagesPath( catalogue["images"].source(), [] )._orderedImages()
+
+	with catalogue.ancestor( Gaffer.ScriptNode ).context() :
+		# There are times when this can be out of sync with the number of images.
+		# Generally when the UI hasn't been opened.
+		currentPlugIndex = min( catalogue["imageIndex"].getValue(), maxIndex )
+
+	catalogueIndex = orderedImages.index( images[currentPlugIndex] )
+	nextIndex = max( min( catalogueIndex + increment, maxIndex ), 0 )
+	nextPlugIndex = images.index( orderedImages[nextIndex] )
+
+	if nextPlugIndex != currentPlugIndex :
+		catalogue["imageIndex"].setValue( nextPlugIndex )
```

Comment:
Should this not be `catalogue["imageIndex"].source()`, so as to support promotion?

Reply:
FIxed in https://github.com/GafferHQ/gaffer/pull/3387/commits/5c110727ad5f1f61ea7d3b0fbaefc50addfeb4e6